File: app/src/main/java/jp/juggler/subwaytooter/StatusButtonsPopup.java
package jp.juggler.subwaytooter;

import android.annotation.SuppressLint;
import android.graphics.drawable.ColorDrawable;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.MotionEventCompat;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.PopupWindow;

import jp.juggler.subwaytooter.api.entity.TootNotification;
import jp.juggler.subwaytooter.api.entity.TootStatusLike;
import jp.juggler.subwaytooter.view.MyListView;

class StatusButtonsPopup {
	
	private final ActMain activity;
	private final View viewRoot;
	private final StatusButtons buttons_for_status;
	
	@SuppressLint("InflateParams")
	StatusButtonsPopup( ActMain activity, Column column, boolean bSimpleList ){
		this.activity = activity;
		this.viewRoot = activity.getLayoutInflater().inflate( R.layout.list_item_popup, null, false );
		this.buttons_for_status = new StatusButtons( activity, column, viewRoot, bSimpleList );
	}
	
	private PopupWindow window;
	
	void dismiss(){
		if( window != null && window.isShowing() ){
			window.dismiss();
		}
	}
	
	@SuppressLint("RtlHardcoded")
	void show(
		@NonNull final MyListView listView
		, @NonNull View anchor
		, @NonNull TootStatusLike status
		, @Nullable TootNotification notification
	){
		
		//
		window = new PopupWindow( activity );
		window.setWidth( WindowManager.LayoutParams.WRAP_CONTENT );
		window.setHeight( WindowManager.LayoutParams.WRAP_CONTENT );
		window.setContentView( viewRoot );
		window.setBackgroundDrawable( new ColorDrawable( 0x00000000 ) );
		window.setTouchable( true );
		window.setOutsideTouchable( true );
		window.setTouchInterceptor( new View.OnTouchListener() {
			@Override public boolean onTouch( View v, MotionEvent event ){
				if( MotionEventCompat.getActionMasked( event ) == MotionEvent.ACTION_OUTSIDE ){
					window.dismiss();
					listView.last_popup_close = SystemClock.elapsedRealtime();
					return true;
				}
				return false;
			}
		} );
		
		buttons_for_status.bind( status, notification );
		buttons_for_status.close_window = window;
		
		int[] location = new int[ 2 ];
		
		anchor.getLocationOnScreen( location );
		int anchor_left = location[ 0 ];
		int anchor_top = location[ 1 ];
		
		listView.getLocationOnScreen( location );
		int listView_top = location[ 1 ];
		
		float density = activity.density;
		
		int clip_top = listView_top + (int) ( 0.5f + 8f * density );
		int clip_bottom = listView_top + listView.getHeight() - (int) ( 0.5f + 8f * density );
		
		int popup_height = (int) ( 0.5f + ( 56f + 24f ) * density );
		int popup_y = anchor_top + anchor.getHeight() / 2;
		
		if( popup_y < clip_top ){
			// 画面外のは画面内にする
			popup_y = clip_top;
		}else if( clip_bottom - popup_y < popup_height ){
			// 画面外のは画面内にする
			if( popup_y > clip_bottom ) popup_y = clip_bottom;
			
			// 画面の下側にあるならポップアップの吹き出しが下から出ているように見せる
			viewRoot.findViewById( R.id.ivTriangleTop ).setVisibility( View.GONE );
			viewRoot.findViewById( R.id.ivTriangleBottom ).setVisibility( View.VISIBLE );
			popup_y -= popup_height;
		}
		
		int anchor_width = anchor.getWidth();
		int popup_width = getViewWidth( viewRoot );
		int popup_x = anchor_left + anchor_width / 2 - popup_width / 2;
		if( popup_x < 0 ) popup_x = 0;
		int popup_x_max = activity.getResources().getDisplayMetrics().widthPixels - popup_width;
		if( popup_x > popup_x_max ) popup_x = popup_x_max;
		
		window.showAtLocation(
			listView
			, Gravity.LEFT | Gravity.TOP
			, popup_x
			, popup_y
		);
	}
	
	private static int getViewWidth( View v ){
		int spec = View.MeasureSpec.makeMeasureSpec( 0, View.MeasureSpec.UNSPECIFIED );
		v.measure( spec, spec );
		return v.getMeasuredWidth();
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/StatusButtonsPopup.kt
package jp.juggler.subwaytooter

import android.annotation.SuppressLint
import android.graphics.drawable.ColorDrawable
import android.os.SystemClock
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.PopupWindow
import androidx.recyclerview.widget.RecyclerView
import jp.juggler.subwaytooter.api.entity.TootNotification
import jp.juggler.subwaytooter.api.entity.TootStatus
import jp.juggler.util.LogCategory
import org.jetbrains.anko.matchParent

internal class StatusButtonsPopup(
    private val activity: ActMain,
    column: Column,
    bSimpleList: Boolean,
    itemViewHolder: ItemViewHolder,
) {

    companion object {

        @Suppress("unused")
        private var log = LogCategory("StatusButtonsPopup")

        private fun getViewWidth(v: View): Int {
            val spec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
            v.measure(spec, spec)
            return v.measuredWidth
        }

        var lastPopupClose = 0L
    }

    private val viewRoot: View
    private val buttonsForStatus: StatusButtons
    private var window: PopupWindow? = null

    init {
        @SuppressLint("InflateParams")
        this.viewRoot = activity.layoutInflater.inflate(R.layout.list_item_popup, null, false)
        val statusButtonsViewHolder = StatusButtonsViewHolder(activity, matchParent, 0f)
        viewRoot.findViewById<LinearLayout>(R.id.llBarPlaceHolder)
            .addView(statusButtonsViewHolder.viewRoot)
        this.buttonsForStatus = StatusButtons(
            activity,
            column,
            bSimpleList,
            statusButtonsViewHolder,
            itemViewHolder
        )
    }

    fun dismiss() {
        val window = this.window
        if (window != null && window.isShowing) {
            window.dismiss()
        }
    }

    @SuppressLint("RtlHardcoded", "ClickableViewAccessibility")
    fun show(
        listView: RecyclerView,
        anchor: View,
        status: TootStatus,
        notification: TootNotification?,
    ) {

        val window = PopupWindow(activity)
        this.window = window

        window.width = WindowManager.LayoutParams.WRAP_CONTENT
        window.height = WindowManager.LayoutParams.WRAP_CONTENT
        window.contentView = viewRoot
        window.setBackgroundDrawable(ColorDrawable(0x00000000))
        window.isTouchable = true
        window.isOutsideTouchable = true
        window.setTouchInterceptor { _, event ->
            if (event.action == MotionEvent.ACTION_OUTSIDE) {
                // ポップアップの外側をタッチしたらポップアップを閉じる
                // また、そのタッチイベントがlistViewに影響しないようにする
                window.dismiss()
                lastPopupClose = SystemClock.elapsedRealtime()
                true
            } else {
                false
            }
        }

        buttonsForStatus.bind(status, notification)
        buttonsForStatus.closeWindow = window

        val location = IntArray(2)

        anchor.getLocationOnScreen(location)
        val anchorLeft = location[0]
        val anchorTop = location[1]

        listView.getLocationOnScreen(location)
        val listviewTop = location[1]

        val density = activity.density

        val clipTop = listviewTop + (0.5f + 8f * density).toInt()
        val clipBottom = listviewTop + listView.height - (0.5f + 8f * density).toInt()

        val popupHeight = (0.5f + (56f + 24f) * density).toInt()
        var popupY = anchorTop + anchor.height / 2

        if (popupY < clipTop) {
            // 画面外のは画面内にする
            popupY = clipTop
        } else if (clipBottom - popupY < popupHeight) {
            // 画面外のは画面内にする
            if (popupY > clipBottom) popupY = clipBottom

            // 画面の下側にあるならポップアップの吹き出しが下から出ているように見せる
            viewRoot.findViewById<View>(R.id.ivTriangleTop).visibility = View.GONE
            viewRoot.findViewById<View>(R.id.ivTriangleBottom).visibility = View.VISIBLE
            popupY -= popupHeight
        }

        val anchorWidth = anchor.width
        val popupWidth = getViewWidth(viewRoot)
        var popupX = anchorLeft + anchorWidth / 2 - popupWidth / 2
        if (popupX < 0) popupX = 0
        val popupXMax = activity.resources.displayMetrics.widthPixels - popupWidth
        if (popupX > popupXMax) popupX = popupXMax

        window.showAtLocation(listView, Gravity.LEFT or Gravity.TOP, popupX, popupY)
    }
}
